state_machine.py

```diff
@@ -103,9 +98,21 @@ def update(self):
                 self.state.update(self)
             except Exception as e:
                 log.exception('Uncaught exception while processing state %s: %s', self.state, str(e))
-                self.go_to_state('error')
-        # CHECK: if there is no current state set (e.g. faulty update() function in some state),
-        # this silently returns without errors. Throwing an exception here might be better.
+                if 'error' in self.states:
+                    self.go_to_state('error')
+                else:
+                    raise Exception("Faulty state {}".format(repr(e)))
+
+    def concat_state_log(self):
```

Comment:
Should this function be defined inside of the state machine?

Reply:
Since this function operates on stateMachine variables, and for the sake of encapsulation I would keep it in the state-machine